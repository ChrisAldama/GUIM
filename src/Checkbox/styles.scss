@import '../styles/colors.scss';

.GUIMCheckbox {
  -webkit-appearance: none;
  background-color: var(--white);
  border: 2px solid var(--medium-gray);
  border-radius: 3px;
  display: inline-block;
  position: relative;
  outline: none;
  width: 16px;
  height: 16px;
  transition: all 0.5s ease;
  cursor: pointer;
}
.GUIMCheckbox:checked{
  background-color: var(--main-blue);
  border: 2px solid var(--main-blue);
  &::before{
    -webkit-transform: rotate(45deg);
    transform: rotate(45deg);
    content: "";
    position: absolute;
    top: -1px;
    left: 4px;
    display: table;
    width: 2px;
    height: 10px;
    border: 2px solid #fff;
    border-top-width: 0;
    border-left-width: 0;
  }
}
