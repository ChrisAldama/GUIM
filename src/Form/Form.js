import React, {
  Component,
  Children,
  isValidElement
} from "react";

const extractContent = child => {
  let response = {};
  switch (child.props.guimInput) {
    case "checkbox":
      response[child.props.name] = child.props.checked;
      break;
    case "picker":
      response[child.props.name] = child.props.active;
      break;
    case "slider":
      response[child.props.name] = child.props.selected_range;
      break;
  }
  return response;
}


class Form extends Component {
  constructor(props) {
    super(props);
    this.grabFormData = this.grabFormData.bind(this);
  }

  grabFormData() {
    let formData = {};

    const traverseChildren = childrenTraverse => {
      Children.forEach(childrenTraverse, childTraverse => {
        if (isValidElement(childTraverse)) {
          traverseChildren(childTraverse.props.children);
          if (childTraverse.props.hasOwnProperty("guimInput")) {
            formData = {...formData, ...extractContent(childTraverse)};
          }
        }
      });
    }

    traverseChildren(this.props.children);
    return formData;
  }

  render() {
    return (
      <div
        className={this.props.className}
        style={this.props.styles}
        name={this.props.name} >
        {this.props.children}
      </div>
    );
  }
}

Form.defaultProps = {
  className: "GUIMForm",
  styles: {},
  name: "form",
  guimInput: "form"
}

export default Form;
