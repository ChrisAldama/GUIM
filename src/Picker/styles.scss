@import '../styles/colors.scss';
@import '../styles/themes.scss';

.GUIMPicker {
  display: flex;
  align-items: center;
  text-align: center;
  cursor: pointer;
  height: 30px;
  & a {
    border: 1px solid var(--main-gray);
    color: var(--dark-gray);
    display: flex;
    background-color: var(--white);
    width: calc(51% - 1px);
    justify-content: center;
    align-items: center;
    transition: all 0.3s ease;
    height: 100%;
    &:last-child{
      border-top-right-radius: 3px;
      border-bottom-right-radius: 3px;
    }
    &:first-child{
      border-top-left-radius: 3px;
      border-bottom-left-radius: 3px;
    }
  }
}

.GUIMPickerBlue {
  & a:hover {
    @apply --blue-theme-hover;
  }
  & .active {
    @apply --blue-theme;
  }
}
.GUIMPickerGray {
  & a:hover {
    @apply --gray-theme-hover;
  }
  & .active {
    @apply --gray-theme;
  }
}
.GUIMPickerGreen {
  & a:hover {
    @apply --green-theme-hover;
  }
  & .active {
    @apply --green-theme;
  }
}
