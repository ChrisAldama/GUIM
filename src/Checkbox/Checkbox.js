import React from 'react';
import './styles.scss';

const Checkbox = props => (
  <input
    className={props.className}
    style={props.styles}
    type="checkbox"
    name={props.name}
    onChange={props.onChange}
    checked={props.checked} />
);

Checkbox.defaultProps = {
  name: 'checkbox',
  onChange: () => console.log('Checkbox onChange()'),
  checked: false,
  styles: {},
  className: 'GUIMCheckbox',
  guimInput: 'checkbox'
};

export default Checkbox;
