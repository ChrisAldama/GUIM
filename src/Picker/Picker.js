import React from 'react';
import './styles.scss';

const Pill = props => (
  <a
    onClick={props.select}
    className={props.active ? 'active' : ''}
    key={props.value}
  >
    {props.label}
  </a>
);

const Picker = props => (
  <div
    name={props.name}
    className={`GUIMPicker ${props.className} ${themes[props.color] || themes.blue}`}
  >
    {
      props.options.map(e =>
        (<Pill
          key={e.value}
          select={props.onChange.bind(null, e.label, e.value)}
          active={props.active.indexOf(e.value) !== -1}
          value={e.value}
          label={e.label}
        />))
    }
  </div>
);

Picker.defaultProps = {
  name: 'picker',
  onChange: (label, value) => console.log(`Picker:active => label: ${label}, value: ${value}`),
  options: [
    { label: 'Uno', value: 1 },
    { label: 'Dos', value: 2 },
  ],
  className: '',
  color: 'blue',
  guimInput: 'picker',
  active: [1],
};

const themes = {
  blue: 'GUIMPickerBlue',
  gray: 'GUIMPickerGray',
  green: 'GUIMPickerGreen',
};

export default Picker;
