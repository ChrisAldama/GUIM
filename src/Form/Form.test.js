import React from 'react';
import Form from './Form';
import Checkbox from '../Checkbox';
import Picker from '../Picker';

import renderer from 'react-test-renderer';

import { mount, configure } from 'enzyme';
import enzymeSerializer from 'enzyme-to-json/serializer';
import Adapter from 'enzyme-adapter-react-16';

expect.addSnapshotSerializer(enzymeSerializer);
configure({ adapter: new Adapter() });

describe('<Form />', () => {
  let active = 1;
  let checked = false;
  let component;
  let form;

  const reRender = () => {
    component = renderer.create(
      <Form ref={ref => (form = ref)} >
        <Checkbox
          name="checkbox"
          checked={checked}
          onChange={() => checked = !checked} />
        <Picker
          onChange={(label, value) => active = value}
          options={[
            {label: "p1", value: 1},
            {label: "p2", value: 2}
          ]}
          active={active}
        />
      </Form>)
  };

  it('match dom snapshot', () => {
    component = mount(
      <Form ref={ref => (form = ref)} >
        <Checkbox
          name="checkbox"
          checked={checked}
          onChange={() => checked = !checked} />
        <Picker
          onChange={(label, value) => active = value}
          options={[
            {label: "p1", value: 1},
            {label: "p2", value: 2}
          ]}
          active={active}
        />
      </Form>);

    expect(component).toMatchSnapshot();
  });

  it('returns children data', () => {
    reRender();
    // Grab data for the first time
    expect(form.grabFormData()).toMatchSnapshot();
  });

  it('returns children data updated after click', () => {
    // click on checkbox to change data
    let tree = component.toJSON();
    tree.children[0].props.onChange();

    reRender();
    expect(form.grabFormData()).toMatchSnapshot();
  });
});
