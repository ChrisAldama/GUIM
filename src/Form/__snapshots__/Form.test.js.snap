// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> match dom snapshot 1`] = `
<Form
  className="GUIMForm"
  guimInput="form"
  name="form"
  styles={Object {}}
>
  <div
    className="GUIMForm"
    name="form"
    style={Object {}}
  >
    <Checkbox
      checked={false}
      className="GUIMCheckbox"
      guimInput="checkbox"
      name="checkbox"
      onChange={[Function]}
      styles={Object {}}
    >
      <input
        checked={false}
        className="GUIMCheckbox"
        name="checkbox"
        onChange={[Function]}
        style={Object {}}
        type="checkbox"
      />
    </Checkbox>
    <Picker
      active={1}
      className="GUIMPicker"
      guimInput="picker"
      name="picker"
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "p1",
            "value": 1,
          },
          Object {
            "label": "p2",
            "value": 2,
          },
        ]
      }
    >
      <div
        className="GUIMPicker"
        name="picker"
      >
        <Pill
          active={true}
          key="1"
          label="p1"
          select={[Function]}
          value={1}
        >
          <a
            className="active"
            key="1"
            onClick={[Function]}
          >
            p1
          </a>
        </Pill>
        <Pill
          active={false}
          key="2"
          label="p2"
          select={[Function]}
          value={2}
        >
          <a
            className=""
            key="2"
            onClick={[Function]}
          >
            p2
          </a>
        </Pill>
      </div>
    </Picker>
  </div>
</Form>
`;

exports[`<Form /> returns children data 1`] = `
Object {
  "checkbox": false,
  "picker": 1,
}
`;

exports[`<Form /> returns children data updated after click 1`] = `
Object {
  "checkbox": true,
  "picker": 1,
}
`;
